src/query/sql/src/planner/optimizer/property/enforcer.rs

```diff
@@ -108,3 +115,17 @@ pub fn enforce_distribution(distribution: &Distribution, s_expr: &SExpr) -> Resu
         )),
     }
 }
+
+fn check_merge(s_expr: &SExpr) -> bool {
```

Comment:
I think what you need is checking if the derived distribution of a `SExpr` is `Distribution::Serial`.

Reply:
In fact, I want to check if union's child plan has `Exchange::Merge`, if has, there will no `ExchangeSource`.So we don't make union shuffle.